### DISCUSSION TITLE
Add option port to salt-config.sh for macos

### DISCUSSION THREAD

**Discussion by HectorValman** [saltstack/salt#67248] [Feature Requests]
Hi

I think it is necessary to add the port option to the salt-config.sh configurator for macOS.

I did it for my needs by adding it to the script: 

```
#!/bin/bash
############################################################################
# Commandline Help
############################################################################
display_help() {
    echo "################################################################"
    echo "Salt Minion Configuration Script"
    echo
    echo "Use this script to configure the minion id as well as the master"
    echo "the minion should connect to. The settings will be changed and"
    echo "the service will be restarted. Must be run as sudo"
    echo
    echo "This script accepts the following parameters:"
    echo
    echo "  -i, --minion-id    The ID to assign this minion"
    echo "  -m, --master       The hostname/IP address of the master"
    echo "  -p, --minion_port       The port of the master"
    echo "  -h, --help         Display this help message"
    echo
    echo "Examples:"
    echo
    echo "  sudo salt-config -i mac_minion -m master.apple.com"
    echo
    echo "  sudo salt-config --minion-id mac_minion --master 10.10.1.10"
    echo
    echo "################################################################"
    exit 1
}

############################################################################
# Parameters
############################################################################
# Initialize Parameters
master=''
minion_id=''
minion_port=''
changed=0

############################################################################
# Check for parameters
############################################################################
# Check for no parameters
if [ $# -eq 0 ] ; then
    echo "ERROR: No Parameters Passed"
    echo "       To see help use --help"
    exit 1
fi

# Check for valid parameters
while [ $# -gt 0 ]; do
    case "$1" in
        -i | --minion-id)
            minion_id="$2"
            shift 2
            ;;
        -m | --master)
            master="$2"
            shift 2
            ;;
        -p | --port)
            minion_port="$2"
            shift 2
            ;;
        -h | --help) # Display Help
            display_help
            ;;
        *)
            break
    esac
done

# Check for additional parameters
if [ ! -z "$1" ] ; then
    echo "ERROR: Unknown Parameter Passed: $1"
    echo "       To see help use --help"
    exit 1
fi

############################################################################
# minion.d directory
############################################################################
if [ ! -d "/etc/salt/minion.d" ]; then
    mkdir /etc/salt/minion.d
fi

############################################################################
# Minion ID
############################################################################
if [ ! -z "$minion_id" ]; then
    echo "Changing minion ID: $minion_id"
    sed -i '' -e '/id:/ s/^#*/#/' /etc/salt/minion
    echo "id: $minion_id" > /etc/salt/minion.d/minion_id.conf
    changed=1
fi

############################################################################
# Master ID
############################################################################
if [ ! -z "$master" ]; then
    echo "Changing master: $master"
    sed -i '' -e '/master:/ s/^#*/#/' /etc/salt/minion
    echo "master: $master" > /etc/salt/minion.d/master_id.conf
    changed=1
fi


############################################################################
# PORT ID
############################################################################
if [ ! -z "$master" ]; then
    echo "Changing post: $minion_port"
    sed -i '' -e '/port:/ s/^#*/#/' /etc/salt/minion
    echo "master_port: $minion_port" > /etc/salt/minion.d/port_id.conf
    changed=1
fi

############################################################################
# Restart Minion
############################################################################
if (( changed == 1 )); then
    echo "Restarting the minion service..."
    launchctl kickstart -k system/com.saltstack.salt.minion
fi
exit 0
```

Best regards

**Comment by welcome** [comment]
Hi there! Welcome to the Salt Community! Thank you for making your first contribution. We have a lengthy process for issues and PRs. Someone from the Core Team will follow up as soon as possible. In the meantime, here’s some information that may help as you continue your Salt journey.
Please be sure to review our [Code of Conduct](https://github.com/saltstack/salt/blob/master/CODE_OF_CONDUCT.md). Also, check out some of our community resources including:

  - [Community Wiki](https://github.com/saltstack/community/wiki)
  - [Salt’s Contributor Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html)
  - [Join our Community Slack](https://join.slack.com/t/saltstackcommunity/shared_invite/zt-3av8jjyf-oBQ2M0vhXOhJpNpRkPWBvg)
  - [IRC on LiberaChat](https://web.libera.chat/#salt)
  - [Salt Project YouTube channel](https://www.youtube.com/channel/UCpveTIucFx9ljGelW63-BWg)
  - [Salt Project Twitch channel](https://www.twitch.tv/saltprojectoss)

There are lots of ways to get involved in our community. Every month, there are around a dozen opportunities to meet with other contributors and the Salt Core team and collaborate in real time. The best way to keep track is by subscribing to the Salt Community Events Calendar.
If you have additional questions, email us at saltproject@vmware.com. We’re glad you’ve joined our community and look forward to doing awesome things with you!

**Comment by OrangeDog** [comment]
It should be `port`, not `minion_port`.

**Comment by rrrix** [comment]
Hi @HectorValman, can you elaborate on why this specific option is needed - what is your use case, and do you currently have any workarounds?

I think this feature request is not something this particular script should support. I personally disagree with this request - why is this configuration option necessary (versus any other arbitrary option?) 

Neither `minion_port` nor `port` are valid configuration options - which "port" configuration option does this specify - the [`master_port`](https://docs.saltproject.io/en/latest/ref/configuration/minion.html#master-port) or [`publish_port`](https://docs.saltproject.io/en/latest/ref/configuration/minion.html#publish-port)?

Also, the [_**master:port**_](https://docs.saltproject.io/en/latest/ref/configuration/minion.html#master-port-syntax) configuration syntax is already available, which makes this request already possible, e.g.: 
```text
# Same as "master_port: 4506"
% /opt/salt/bin/salt-config.sh --master master_ip.example.com:4506
```

> ### [MASTER:PORT SYNTAX](https://docs.saltproject.io/en/latest/ref/configuration/minion.html#master-port-syntax)
> __New in version 2015.8.0.__
> 
> The `master` config option can also be set to use the master's IP in conjunction with a port number by default.
> 
> ```yaml
> master: localhost:1234
> ```
> For IPv6 formatting with a port, remember to add brackets around the IP address before adding the port and enclose the line in single quotes to make it a string:
> ```yaml
> master: '[2001:db8:85a3:8d3:1319:8a2e:370:7348]:1234'
> ```
> > #### Note
> > If a port is specified in the master as well as master_port, the master_port setting will be overridden by the master configuration.

> ### [MASTER_PORT](https://docs.saltproject.io/en/latest/ref/configuration/minion.html#master-port)
> Default: `4506`
> The port of the master ret server, this needs to coincide with the `ret_port` option on the Salt master.
> ```yaml
> master_port: 4506
> ```
> ### [PUBLISH_PORT](https://docs.saltproject.io/en/latest/ref/configuration/minion.html#publish-port)
> Default: `4505`
> The port of the master publish server, this needs to coincide with the `publish_port` option on the Salt master.
> ```yaml
> publish_port: 4505
> ```

Lastly, the ubiquitous [salt-bootstrap.sh](https://github.com/saltstack/salt-bootstrap/blob/develop/bootstrap-salt.sh#L328-L415) script doesn't even have this option - instead it supports the `-j` flag for dynamic configuration during bootstrapping:

```text
-j  Replace the Minion config file with data passed in as a JSON string. If
    a Minion config file is found, a reasonable effort will be made to save
    the file with a ".bak" extension. If used in conjunction with -C or -F,
    no ".bak" file will be created as either of those options will force
    a complete overwrite of the file.
```

I think instead `salt-config.sh` could re-use this same functionality to support dynamic minion configuration for _any_ option (not just the ambiguous `port`, whichever option may imply). 

Lastly, I've never personally understood why this script even exists, as it's unique to macOS, and it's usually both trivial and necessary to write ad-hoc configuration to `/etc/salt/minion.d/` before starting the minion.
